saas/models/saas_template.py

```diff
@@ -0,0 +1,144 @@
+        'Modules to install',
+        help='Domain to search for modules to install after template database creation',
+        defalt="[]")
+    template_post_init = fields.Text(
+        'Template Initialization',
+        default=lambda s: s.env['ir.actions.server'].DEFAULT_PYTHON_CODE,
+        help='Python code to be executed once db is created and modules are installed')
+    operator_ids = fields.One2many(
+        'saas.template.operator',
+        'template_id')
+
+    @api.constrains('code')
+    def _check_python_code(self):
+        for r in self.sudo():
+            msg = test_python_expr(expr=r.code.strip(), mode="exec")
+            if msg:
+                raise ValidationError(msg)
+
+
+class SAASTemplateLine(models.Model):
+    _name = 'saas.template.operator'
+    _description = 'Template\'s Settings for Operator'
+
+    template_id = fields.Many2one('saas.template')
+    operator_id = fields.Many2one('saas.operator')
+    password = fields.Char('DB Password')
+    db_name = fields.Char(required=True)
+    db_id = fields.Many2one('saas.db', readonly=True)
+    db_state = fields.Selection(related='db_id.state')
+    to_rebuild = fields.Boolean()
+    state = fields.Selection([
+        ('draft', 'Draft'),
+        ('creating', 'Database Creating'),
+        ('installing_modules', 'Modules installation'),
+        ('post_init', 'Extra initialization'),
+        ('done', 'Ready'),
+
+    ])
+
+    def preparing_template_next(self):
+        # TODO: This method is called by cron every few minutes
+        template_operators = self.search([('to_rebuild', '=', True)])
+        operators = template_operators.mapped('operator_id')
+
+        # filter out operators which already have template creating
+        def filter_free_operators(op):
+            states = op.template_operator_ids.mapped('state')
+            return all((s in ['draft', 'done'] for s in states))
+
+        operators = operators.filtered(filter_free_operators)
+        if not operators:
+            # it's not a time to start
+            return
+
+        for t_op in template_operators:
+            if t_op.operator_id not in operators:
+                continue
+            t_op._prepare_template()
+
+            # only one template per operator
+            operators -= t_op.operator_id
+
+    def _prepare_template(self):
+        for r in self:
+            # delete db is there is one
+            r.db_id.drop_db()
+            if not r.db_id or r.operator_id != r.db_id.operator_id:
+                r.db_id = self.env['saas.db'].create({
+                    'name': r.db_name,
+                    'operator_id': r.operator_id,
+                    'type': 'template',
+                })
+            password = random_password()
+            self.env['saas.log'].log_db_creating(self)
+
+            r.write({
+                'state': 'creating',
+                'password': password,
+            })
+
+            r.db_id.with_delay().create_db(
```

Comment:
AttributeError: 'saas.db' object has no attribute 'with_delay'

Reply:
Try to add ``@job`` decorator

Reply:
Actually, it's there. Check OCA/queue for configuration: https://github.com/OCA/queue/tree/11.0/queue_job#id13

Try to use this module to confirm configuration: https://github.com/OCA/queue/tree/11.0/base_import_async

For test environment check for documentation about how this module should be run: https://github.com/OCA/queue/tree/11.0/test_queue_job

Reply:
base_import_async is uninstallable on odoo 12